server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java

```diff
@@ -0,0 +1,171 @@
+        hasBeenShutdown = true;
+        final var maxTimeout = Node.MAXIMUM_SHUTDOWN_TIMEOUT_SETTING.get(settings);
+        final var reindexTimeout = Node.MAXIMUM_REINDEXING_TIMEOUT_SETTING.get(settings);
+
+        record Stopper(String name, SubscribableListener<Void> listener) {
+            boolean isIncomplete() {
+                return listener().isDone() == false;
+            }
+        }
+
+        final var stoppers = new ArrayList<Stopper>();
+        final var allStoppersFuture = new PlainActionFuture<Void>();
+        try (var listeners = new RefCountingListener(allStoppersFuture)) {
+            final BiConsumer<String, Runnable> stopperRunner = (name, action) -> {
+                final var stopper = new Stopper(name, new SubscribableListener<>());
+                stoppers.add(stopper);
+                stopper.listener().addListener(listeners.acquire());
+                new Thread(() -> {
+                    try {
+                        action.run();
+                    } catch (Exception ex) {
+                        logger.warn("unexpected exception in shutdown task [" + stopper.name() + "]", ex);
+                    } finally {
+                        stopper.listener().onResponse(null);
+                    }
+                }, stopper.name()).start();
+            };
+
+            stopperRunner.accept("http-server-transport-stop", httpServerTransport::close);
+            stopperRunner.accept("async-search-stop", () -> awaitSearchTasksComplete(maxTimeout, taskManager));
+            stopperRunner.accept("reindex-stop", () -> awaitReindexTasksComplete(reindexTimeout, taskManager));
+            if (terminationHandler != null) {
+                stopperRunner.accept("termination-handler-stop", terminationHandler::handleTermination);
+            }
+        }
+
+        final Supplier<String> incompleteStoppersDescriber = () -> stoppers.stream()
+            .filter(Stopper::isIncomplete)
+            .map(Stopper::name)
+            .collect(Collectors.joining(", ", "[", "]"));
+
+        try {
+            if (TimeValue.ZERO.equals(maxTimeout)) {
+                allStoppersFuture.get();
+            } else {
+                allStoppersFuture.get(maxTimeout.millis(), TimeUnit.MILLISECONDS);
+            }
+        } catch (ExecutionException e) {
+            assert false : e; // listeners are never completed exceptionally
+            logger.warn("failed during graceful shutdown tasks", e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            logger.warn("interrupted while waiting for graceful shutdown tasks: " + incompleteStoppersDescriber.get(), e);
+        } catch (TimeoutException e) {
+            logger.warn("timed out while waiting for graceful shutdown tasks: " + incompleteStoppersDescriber.get());
+        }
+    }
+
+    private void waitForTasks(TimeValue timeout, String taskName, TaskManager taskManager) {
```

Comment:
Looks like we have differing naming for similar things in this class, I'd prefer to name this: `awaitTasksComplete`, similar to the two other `await` methods.

Reply:
Yes, that's a better name. Thanks.